packages/composite-checkout/src/components/checkout-modal.js

```diff
@@ -0,0 +1,169 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect } from 'react';
+import styled, { keyframes } from 'styled-components';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import joinClasses from '../lib/join-classes';
+import Button from './button';
+import { useLocalize } from '../lib/localize';
+
+export default function CheckoutModal( {
+	className,
+	title,
+	copy,
+	primaryAction,
+	closeModal,
+	isVisible,
+	buttonCTA,
+	cancelButtonCTA,
+} ) {
+	const localize = useLocalize();
+	useEffect( () => {
+		if ( isVisible ) {
+			document.body.style = 'overflow: hidden';
+			document.addEventListener(
+				'keydown',
+				key => {
+					handleKeyPress( key, closeModal );
+				},
+				false
+			);
+
+			return;
+		}
+
+		document.body.style = 'overflow: scroll';
+	} );
```

Comment:
Without the second argument to `useEffect` ([the dependencies array](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)), this will run after every render, which is not terrible but I think it only needs to occur when `isVisible` changes. I'd try modifying this to be as follows (I also put it into a custom hook to make it easier to read).

The return value of `useEffect` is a function to call when the component is unmounted or when its dependencies are changed.

```js
function useModalScreen( isVisible, closeModal ) {
  useEffect( () => {
    document.body.style = isVisible ? 'overflow: hidden' : 'overflow: scroll';
    const keyPressHandler = makeHandleKeyPress( closeModal );
    if ( isVisible ) {
      document.addEventListener( 'keydown', keyPressHandler, false );
    }
    return () => document.removeEventListener( 'keydown', keyPressHandler, false );
  }, [ isVisible, closeModal ] );
}

function makeHandleKeyPress( closeModal ) {
  const escapeKey = 27;
  return key => {
    if ( key.keyCode === escapeKey ) {
      closeModal();
    }
  };
}
```

Reply:
Thanks for this. I noticed in the coupon field PR, where I am doing something similar, that it was calling one of the methods multiple times but wasn't sure why. It wasn't going terribly out of control so I thought it was ok. I'll see if I can apply this approach there as well.